server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java

```diff
@@ -1215,18 +1245,23 @@ public void parseXContent(XContentParser parser, boolean checkTrailingTokens) th
                 }
             } else if (token == XContentParser.Token.START_OBJECT) {
                 if (QUERY_FIELD.match(currentFieldName, parser.getDeprecationHandler())) {
-                    queryBuilder = parseTopLevelQuery(parser);
+                    queryBuilder = parseTopLevelQuery(parser, searchUsage::trackQueryUsage);
```

Comment:
as for sections, I want for tracking only the ones that make sense. Some like explain and profile did not feel worthwhile to track. I may be tracking too much already.

I wonder if we should track usages of query and knn at the same time through a special section name.

Reply:
Some of these I think are only interesting if they are the non-default value? So _source is only interesting if its disabled, or if filters are applied; stored_fields is only interesting if it's not empty, etc.  Client code might generate queries that include default settings and if we count this I think we'll get misleading information from the telemetry.

Reply:
agreed, I was initially torn on adding extra logic here, but I can see how clients like Kibana may provide e.g. empty runtime mappings all the time, which would make no sense to track. I added some logic to conditionally track and a test that verifies that either parsing fails for empty sections or they are not tracked.

Keep in mind that there's a whole set of errors that may happen after parsing, which won't affect whether we track or not. That means that failed requests will still be tracked, as long as they pass parsing, but I think this is an expected consequence of tracking at parse time, focusing on incoming requests.